### PR TITLE
feat: overhaul logging system with unified paths, rotation, and structured output

### PR DESCRIPTION
**Pull request by AprilNEA**
## Summary

- Replace fragmented logging (4 different paths across Desktop/CLI/launchd modes) with unified daemon-managed system
- New `arcbox-logging` crate: size-based rotation (10MB/file, 5 max), JSON file output + human-readable stderr
- New `abctl logs` CLI command with `--follow`, `--component`, `--lines`
- Helper gets `tracing` integration (replaces `eprintln!`)
- Agent dual-output: hvc1 console (or stderr fallback) + `/arcbox/log/agent.log` via VirtioFS

## Changes by component

| Component | Change |
|-----------|--------|
| `arcbox-logging` (new) | `LogConfig`, `SizeRotatingWriter`, `init()` / `init_with_sentry()` |
| `arcbox-daemon` | Uses `arcbox_logging::init_with_sentry()`, adds `--foreground` flag |
| `arcbox-helper` | Adds `tracing`, writes to `/var/log/arcbox/helper.log` |
| `arcbox-agent` | Dual output: hvc1 console (no ANSI) + `/arcbox/log/agent.log` via VirtioFS, with graceful fallback if log dir creation fails |
| `arcbox-cli` | Removes fd redirection from `daemon start`, new `abctl logs` command, forwards `--foreground` to daemon binary |
| `install.rs` | Plist no longer sets `StandardOutPath`/`StandardErrorPath` |
| `arcbox-constants` | New log file name constants + `privileged_log` module |

## Unified log paths (after this PR)

| Component | Path | Format |
|-----------|------|--------|
| daemon | `~/.arcbox/log/daemon.log` | JSON |
| helper | `/var/log/arcbox/helper.log` | JSON |
| agent | `~/.arcbox/log/agent.log` | text |
| containerd | `~/.arcbox/log/containerd.log` | text |
| dockerd | `~/.arcbox/log/dockerd.log` | text |

## Test plan

- [x] `cargo check --workspace` passes
- [x] `cargo clippy --workspace` passes
- [x] `cargo test -p arcbox-logging` passes (rotation unit tests)
- [x] Run `arcbox-daemon` — verify `~/.arcbox/log/daemon.log` created with JSON output
- [x] Run `arcbox-daemon --foreground` — verify stderr shows human-readable + file gets JSON
- [x] E2E: `docker run --rm alpine echo hello` → `hello`
- [x] E2E: `docker run -p 8080:80 nginx` + `curl localhost:8080` → HTTP 200
- [x] Agent logs written to `~/.arcbox/log/agent.log` via VirtioFS
- [x] `abctl logs` shows daemon log output
- [x] `abctl logs -f` follows new output (verified with docker run trigger)
- [x] `abctl logs --component helper` — correct path, reasonable error when helper not running
- [x] Log rotation unit tests pass (rotation_creates_numbered_files, oldest_file_is_deleted)

**Note**: `arcbox-desktop` plist change (remove `StandardOutPath`/`StandardErrorPath`) is in the desktop repo and should be merged alongside this PR.

### CODE REVIEW

File: bundle/com.arcboxlabs.desktop.helper.plist [hunk 1]
```diff
@@ -6,6 +6,8 @@
     <string>com.arcboxlabs.desktop.helper</string>
     <key>Program</key>
     <string>/usr/local/libexec/arcbox-helper</string>
+    <key>StandardErrorPath</key>
+    <string>/var/log/arcbox/helper.err</string>
```
[the remark this answers is not in the document]
Comment (AprilNEA) [reply bundle/com.arcboxlabs.desktop.helper.plist]: Fixed in 9e94373 — removed StandardErrorPath from helper plist. The helper creates /var/log/arcbox/ itself via arcbox_logging::init().

[the remark this answers is not in the document]
Comment (AprilNEA) [reply bundle/com.arcboxlabs.desktop.helper.plist]: Fixed in 1d69ef2 — removed StandardErrorPath. Same as above.

File: app/arcbox-cli/src/commands/logs.rs [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+
+use anyhow::{Context, Result, bail};
+use clap::{Args, ValueEnum};
+
+/// Arguments for the logs command.
+#[derive(Debug, Args)]
+pub struct LogsArgs {
+    /// Component to show logs for.
+    #[arg(default_value = "daemon")]
+    pub component: LogComponent,
```
[the remark this answers is not in the document]
Comment (AprilNEA) [reply app/arcbox-cli/src/commands/logs.rs:15]: Fixed in 1fc2625 — changed to `--component/-c` flag.

File: app/arcbox-daemon/src/main.rs [hunk 3]
```diff
@@ -61,23 +67,25 @@ fn main() -> Result<()> {
...
-        .with(tracing_subscriber::fmt::layer().with_target(false))
-        .with(sentry::integrations::tracing::layer())
-        .init();
+    let data_dir = startup::resolve_data_dir(args.data_dir.as_ref());
+    let log_guard = arcbox_logging::init_with_sentry(LogConfig {
+        log_dir: data_dir.join(arcbox_constants::paths::host::LOG),
+        file_name: arcbox_constants::paths::host::DAEMON_LOG.to_string(),
+        default_filter: "arcbox=info,arcbox_daemon=info".to_string(),
+        foreground: args.foreground,
+        ..LogConfig::default()
```
[the remark this answers is not in the document]
Comment (AprilNEA) [reply app/arcbox-daemon/src/main.rs:76]: Fixed in ce9b59a — build_daemon_args() now forwards --foreground.

File: app/arcbox-cli/src/commands/logs.rs [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+/// Print the last `n` lines then follow new output.
+async fn tail_follow(path: &Path, n: usize) -> Result<()> {
+    // Print initial tail.
+    tail_lines(path, n)?;
+
+    // Seek to end and poll for new content.
+    let mut file = std::fs::File::open(path)
+        .with_context(|| format!("Failed to open {}", path.display()))?;
+    file.seek(SeekFrom::End(0))?;
+    let mut reader = BufReader::new(file);
```
[the remark this answers is not in the document]
Comment (AprilNEA) [reply app/arcbox-cli/src/commands/logs.rs]: Fixed in 1fc2625 — tail_follow now detects rotation via inode comparison and reopens.

File: app/arcbox-cli/src/commands/logs.rs [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+            }
+            Ok(_) => {
+                print!("{line}");
+            }
+            Err(e) => {
+                bail!("Error reading log file: {e}");
+            }
+        }
+    }
+}
```
[the remark this answers is not in the document]
Comment (AprilNEA) [reply app/arcbox-cli/src/commands/logs.rs:181]: Fixed in 1fc2625 — same as above.

File: app/arcbox-cli/src/commands/logs.rs [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+/// Print the last `n` lines of a file.
+fn tail_lines(path: &Path, n: usize) -> Result<()> {
+    let file = std::fs::File::open(path)
+        .with_context(|| format!("Failed to open {}", path.display()))?;
+    let reader = BufReader::new(file);
+    let lines: Vec<String> = reader.lines().collect::<Result<_, _>>()?;
+    let start = lines.len().saturating_sub(n);
+    for line in &lines[start..] {
+        println!("{line}");
+    }
```
[the remark this answers is not in the document]
Comment (AprilNEA) [reply app/arcbox-cli/src/commands/logs.rs]: Fixed in 1fc2625 — rewrote to scan backwards in 64KB chunks.

File: app/arcbox-cli/src/commands/logs.rs [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+                LogComponent::Daemon => arcbox_constants::paths::host::DAEMON_LOG,
+                LogComponent::Agent => arcbox_constants::paths::host::AGENT_LOG,
+                LogComponent::Dockerd => "dockerd.log",
+                LogComponent::Containerd => "containerd.log",
+                LogComponent::Helper => unreachable!(),
+            };
+            Ok(log_dir.join(file_name))
+        }
+    }
+}
```
[the remark this answers is not in the document]
Comment (AprilNEA) [reply app/arcbox-cli/src/commands/logs.rs:95]: Acknowledged — kept duplicated with a NOTE comment. arcbox-constants has zero deps; adding `dirs` there would break that invariant. Will consolidate when a shared config crate exists.

File: guest/arcbox-agent/src/main.rs [hunk 8]
```diff
@@ -41,13 +41,37 @@ async fn main() -> Result<()> {
...
+            .with(
+                tracing_subscriber::EnvFilter::try_from_default_env()
+                    .unwrap_or_else(|_| "arcbox_agent=info".into()),
+            )
+            .with(tracing_subscriber::fmt::layer().with_writer(std::io::stderr))
+            .with(
+                tracing_subscriber::fmt::layer()
+                    .with_target(true)
+                    .with_writer(non_blocking),
+            )
```
[the remark this answers is not in the document]
Comment (AprilNEA) [reply guest/arcbox-agent/src/main.rs]: Valid — agent runs in guest VM (linux musl) so it can't use arcbox-logging. Docs updated in 9e94373 to clarify agent has no rotation.

File: guest/arcbox-agent/src/main.rs [hunk 9]
```diff
@@ -41,13 +41,37 @@ async fn main() -> Result<()> {
...
+            .init();
+        Some(guard)
```
[the remark this answers is not in the document]
Comment (AprilNEA) [reply guest/arcbox-agent/src/main.rs]: Fixed in c125698 — now falls back to stderr-only logging on mkdir failure.

File: common/arcbox-logging/src/rotating.rs [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+        let (file, current_size) = open_log_file(&path);
+        Self {
+            inner: Mutex::new(RotatingState {
+                path,
+                max_size,
+                max_files,
+                file,
+                current_size,
+            }),
+        }
```
[the remark this answers is not in the document]
Comment (AprilNEA) [reply common/arcbox-logging/src/rotating.rs:49]: Fixed in 1d69ef2 — added assert!(max_files > 0) and assert!(max_size > 0).

File: common/arcbox-logging/src/rotating.rs [hunk 11]
```diff
@@ -0,0 +1,174 @@
...
+    // Shift existing rotated files: .{n-1} → .{n}, ..., .1 → .2
+    for i in (1..state.max_files).rev() {
+        let from = rotated_path(&state.path, i);
+        let to = rotated_path(&state.path, i + 1);
+        let _ = fs::rename(&from, &to);
+    }
+
+    // Move current log to .1
+    let rotated = rotated_path(&state.path, 1);
+    let _ = fs::rename(&state.path, &rotated);
```
[the remark this answers is not in the document]
Comment (AprilNEA) [reply common/arcbox-logging/src/rotating.rs]: Fixed in 1d69ef2 — rotation errors now logged via eprintln (tracing unavailable inside the writer). If rename of current log fails, continues writing to same file.

File: common/arcbox-logging/src/lib.rs [hunk 12]
```diff
@@ -0,0 +1,150 @@
...
+use tracing_appender::non_blocking::WorkerGuard;
+use tracing_subscriber::{layer::SubscriberExt, util::SubscriberInitExt};
+
+pub use rotating::SizeRotatingWriter;
+
+/// Configuration for log initialization.
+pub struct LogConfig {
+    /// Directory to write log files into (e.g. `~/.arcbox/log`).
+    pub log_dir: PathBuf,
+    /// Log file name prefix (e.g. `"daemon"` → `daemon.log`).
```
[the remark this answers is not in the document]
Comment (AprilNEA) [reply common/arcbox-logging/src/lib.rs]: Fixed in 1d69ef2.

File: docs/data-directories.md [hunk 13]
```diff
@@ -0,0 +1,274 @@
...
+|------|---------|---------|
+| `run/daemon.pid` | Daemon PID file (stale-process detection) | daemon |
+| `run/docker.sock` | Docker-compatible API Unix socket | daemon |
+| `run/arcbox.sock` | gRPC API Unix socket | daemon |
+
+### 1.2 `log/` — Logs
+
+Each component writes its own log file via `tracing-appender` with size-based
+rotation (10 MB per file, 5 rotated files max). Daemon and helper logs are
+JSON; guest logs are plain text.
```
[the remark this answers is not in the document]
Comment (AprilNEA) [reply docs/data-directories.md]: Fixed in 9e94373 — updated to clarify only daemon/helper have size-based rotation.

File: docs/data-directories.md [hunk 14]
```diff
@@ -0,0 +1,274 @@
...
+
+---
+
+## 10. Temporary Paths
+
+| Path | Purpose | Creator |
+|------|---------|---------|
+| `/tmp/arcbox-daemon.stdout.log` | Daemon stdout (Desktop LaunchAgent plist) | launchd |
+| `/tmp/arcbox-daemon.stderr.log` | Daemon stderr (Desktop LaunchAgent plist) | launchd |
+
```
[the remark this answers is not in the document]
Comment (AprilNEA) [reply docs/data-directories.md]: Fixed in 9e94373 — section renamed to 'Legacy Paths' with all old paths listed.

File: app/arcbox-helper/src/server.rs [hunk 15]
```diff
@@ -58,7 +58,7 @@ async fn idle_watchdog(tracker: Arc<ConnectionTracker>) {
         if tracker.is_idle() {
             let start = *idle_since.get_or_insert_with(tokio::time::Instant::now);
             if start.elapsed() >= IDLE_TIMEOUT {
-                eprintln!("arcbox-helper: idle timeout, exiting");
+                tracing::info!("idle timeout, exiting");
                 std::process::exit(0);
```
[the remark this answers is not in the document]
Comment (AprilNEA) [reply app/arcbox-helper/src/server.rs]: Fixed in 7de7f70 — replaced process::exit with CancellationToken-based graceful shutdown so the log guard flushes.

File: docs/data-directories.md [hunk 16]
```diff
@@ -0,0 +1,280 @@
...
+Defined in `common/arcbox-constants/src/paths.rs`, `privileged` module.
+
+| Path | Purpose | Creator |
+|------|---------|---------|
+| `/usr/local/libexec/arcbox-helper` | Privileged helper binary | cli (`abctl _install`) |
+| `/Library/LaunchDaemons/com.arcboxlabs.desktop.helper.plist` | Helper LaunchDaemon (socket-activation) | cli (`abctl _install`) |
+| `/var/run/arcbox-helper.sock` | Helper socket (created by launchd) | launchd |
+| `/var/run/docker.sock` | Symlink → `~/.arcbox/run/docker.sock` | helper |
+| `/etc/resolver/arcbox.local` | macOS DNS resolver file | helper |
```
[the remark this answers is not in the document]
Comment (AprilNEA) [reply docs/data-directories.md:138]: Fixed in efd57d0 — added `/var/log/arcbox/` and `/var/log/arcbox/helper.log` to the privileged paths table.

File: docs/data-directories.md [hunk 17]
```diff
@@ -0,0 +1,280 @@
...
+## Component Responsibility Matrix
+
+| Component | Paths Managed |
+|-----------|--------------|
+| **daemon** | `~/.arcbox/{run,log,data,boot,runtime}/`, sockets, PID, boot asset download, bundle seeding |
+| **cli** (`abctl`) | `~/.arcbox/{bin,shell,completions}/`, Docker context, LaunchAgent registration, shell profile injection |
+| **helper** (root) | `/etc/resolver/`, `/usr/local/bin/` symlinks, `/var/run/docker.sock` symlink |
+| **desktop** | SMAppService LaunchAgent registration, `~/.arcbox/run/` directory creation, helper install trigger |
+| **agent** (guest) | `/arcbox/` mount, `/var/lib/{docker,containerd}/`, guest sockets, Btrfs subvolumes |
+| **launchd** | `/var/run/arcbox-helper.sock` (socket-activation), log file redirection |
```
[the remark this answers is not in the document]
Comment (AprilNEA) [reply docs/data-directories.md]: Fixed in efd57d0 — updated to "helper service lifecycle", removed "log file redirection".

File: app/arcbox-cli/src/commands/logs.rs [hunk 18]
```diff
@@ -0,0 +1,208 @@
...
+        file.seek(SeekFrom::Start(scan_pos))?;
+        let mut buf = vec![0u8; read_size as usize];
+        file.read_exact(&mut buf)?;
+
+        for &b in buf.iter().rev() {
+            if b == b'\n' {
+                newlines_found += 1;
+                // n+1 because the last byte of a file is often '\n' itself.
+                if newlines_found > n {
+                    start_offset = file_len - (file_len - scan_pos) + (buf.len() as u64 - buf.iter().rev().position(|&x| x == b'\n').unwrap() as u64);
```
[the remark this answers is not in the document]
Comment (AprilNEA) [reply app/arcbox-cli/src/commands/logs.rs]: Good catch — fixed in efd57d0. Replaced the re-created reverse iterator with `enumerate().rev()` to track the exact index of the newline that triggers `newlines_found > n`. The offset is now `scan_pos + idx + 1`.

File: app/arcbox-cli/src/commands/logs.rs [hunk 19]
```diff
@@ -0,0 +1,208 @@
...
+            break;
+        }
+    }
+
+    // Seek to the computed start and print everything after it.
+    file.seek(SeekFrom::Start(start_offset))?;
+    let reader = BufReader::new(file);
+    for line in reader.lines() {
+        println!("{}", line?);
+    }
```
[the remark this answers is not in the document]
Comment (AprilNEA) [reply app/arcbox-cli/src/commands/logs.rs:139]: Acknowledged. The tail_lines offset bug from the sibling comment is now fixed. Will add unit tests for tail_lines edge cases as a follow-up.

File: app/arcbox-helper/src/server.rs [hunk 20]
```diff
@@ -69,15 +72,26 @@ async fn idle_watchdog(tracker: Arc<ConnectionTracker>) {
...
+                    Err(e) => {
+                        tracing::warn!("accept error: {e}");
+                        continue;
+                    }
+                }
+            }
+            () = shutdown.cancelled() => {
+                return;
             }
         };
```
[the remark this answers is not in the document]
Comment (AprilNEA) [reply app/arcbox-helper/src/server.rs:96]: The spawned task still calls `tracker.disconnect()` in its drop path, so the count stays consistent. The helper is a short-lived socket-activated process (idle 30s → exit, launchd restarts) — precise connection tracking during shutdown isn't load-bearing here. Not addressing in this PR.

File: common/arcbox-logging/src/lib.rs [hunk 21]
```diff
@@ -0,0 +1,150 @@
...
+///
+/// Returns a [`LogGuard`] that **must** be held until shutdown. Dropping
+/// the guard flushes all pending writes to the log file.
+///
+/// # Panics
+///
+/// Panics if the log directory cannot be created.
+pub fn init(config: LogConfig) -> LogGuard {
+    std::fs::create_dir_all(&config.log_dir).expect("failed to create log directory");
+
```
[the remark this answers is not in the document]
Comment (AprilNEA) [reply common/arcbox-logging/src/lib.rs:72]: Intentional fail-fast — documented in the `# Panics` section of `init()`. If the log directory can't be created, the daemon/helper shouldn't start silently without observability. Callers are expected to ensure prerequisites (which `create_dir_all` in `startup/mod.rs` handles).

File: common/arcbox-logging/src/rotating.rs [hunk 22]
```diff
@@ -0,0 +1,193 @@
...
+    PathBuf::from(p)
+}
+
+fn open_log_file(path: &PathBuf) -> (File, u64) {
+    let file = OpenOptions::new()
+        .create(true)
+        .append(true)
+        .open(path)
+        .unwrap_or_else(|e| panic!("failed to open log file {}: {e}", path.display()));
+
```
[the remark this answers is not in the document]
Comment (AprilNEA) [reply common/arcbox-logging/src/rotating.rs]: Same rationale as the `init()` panic above — `open_log_file` is called after `create_dir_all` succeeds, so an open failure indicates a more fundamental issue (disk full, permission revoked). Fail-fast is the right behavior here.
